fix(ivy): prevent binding off-by-one errors in property bindings

benlesh opened the pull request:
- Extracts and documents code that will be common to interpolation instructions
- Ensures that binding indices are updated at the proper time during compilation
- Adds additional tests

Related #30011

NOTE: This is a direct branch off of #30011, just exlcuding the addition of `textInterpolate` instructions.

kara commented on packages/core/src/render3/instructions/interpolation.ts:
```diff
@@ -0,0 +1,289 @@
+/**
```
Can we roll back this file move too? Unrelated to the bug fix.

benlesh committed:
fix(ivy): property bindings use correct indices

- Extracts and documents code that will be common to interpolation instructions
- Ensures that binding indices are updated at the proper time during compilation
- Adds additional tests

Related #30011

kara commented on packages/core/src/render3/instructions/all.ts:
```diff
@@ -34,6 +34,7 @@ export * from './element';
 export * from './element_container';
 export * from './embedded_view';
 export * from './get_current_view';
+export * from './interpolation';
```
I don't think this file exists right now, right?

benlesh replied:
LOL :man_facepalming:

kara reviewed (approved):
LGTM

kara commented:
merge-assistance: global
